program-analysis/echidna/advanced/state-network-forking.md

```diff
@@ -0,0 +1,96 @@
+# On-chain fuzzing with state forking
+
+**Table of contents:**
+
+- [On-chain fuzzing with state forking](#on-chain-fuzzing-with-state-forking)
+  - [Example](#example)
+  - [Corpus and RPC cache](#corpus-and-rpc-cache)
+  - [Coverage and Etherscan integration](#coverage-and-etherscan-integration)
+
+## Example
+
+One of the most anticipated features of Echidna 2.1.0 is the state network forking. This means that Echidna can run starting with an existing blockchain state provided by an external RPC service (Infura, Alchemy, local node, etc). 
+This enables users to speed up the fuzzing setup when using already deployed contracts. For instance:
+
+```solidity
+interface IHevm {
+    function warp(uint256 newTimestamp) external;
+    function roll(uint256 newNumber) external;
+}
+
+interface Compound {
+  function mint() external payable;
+  function balanceOf(address) external view returns (uint256);
+}
+
+contract TestCompoundEthMint {
+  address constant HEVM_ADDRESS = 0x7109709ECfa91a80626fF3989D68f67F5b1DD12D;
+  IHevm hevm = IHevm(HEVM_ADDRESS);
+  Compound comp = Compound(0x4Ddc2D193948926D02f9B1fE9e1daa0718270ED5);
+
+  constructor() {
+    hevm.roll(16771449);  // sets the correct block number
+    hevm.warp(1678131671); // sets the expected timestamp for the block number
+  }
+  
+  function assertNoBalance() public payable {
+    require(comp.balanceOf(address(this)) == 0);
+    comp.mint{value: msg.value}();
+    assert(comp.balanceOf(address(this)) == 0);
+  }
+}
+```
+
+This test will fail if the minting of cETH success and the balance of the contract increases. In order to use this feature, the user needs to specify the RPC endpoint for Echidna to use before running the fuzzing campaign. This requires using the following environment variables: 
+
+```
+export ECHIDNA_RPC_URL=http://.. ECHIDNA_RPC_BLOCK=16771449
```

Comment:
Might be good to show the one-liner version as well (so people don't clutter env-variables and maybe this shouldn't be permanently set)
```
ECHIDNA_RPC_URL=http://.. ECHIDNA_RPC_BLOCK=16771449 echidna compound.sol
```

Reply:
Done